mypy/semanal.py

```diff
@@ -719,9 +719,44 @@ def analyze_class_body(self, defn: ClassDef) -> Iterator[bool]:
                 yield True
                 self.calculate_abstract_status(defn.info)
                 self.setup_type_promotion(defn)
-
+                self.apply_class_plugin_hooks(defn)
                 self.leave_class()
 
+    def apply_class_plugin_hooks(self, defn: ClassDef) -> None:
+        """Apply a plugin hook that may infer a more precise definition for a class."""
+        def get_fullname(expr: Expression) -> Optional[str]:
+            # We support @foo.bar(...) @foo.bar and @bar
+            # TODO: Support IndexExpressions?
```

Comment:
``IndexExpression`` is prohibited in decorators by the Python parser.

Reply:
But not base classes which also uses this.

Reply:
For base classes yes, this is important, since they can be generic.